lib/openstack_query/runners/query_runner.py

```diff
@@ -0,0 +1,97 @@
+from abc import abstractmethod
+
+from typing import Optional, List, Any, Callable
+
+from openstack_api.openstack_wrapper_base import OpenstackWrapperBase
+from openstack_api.openstack_connection import OpenstackConnection
+from custom_types.openstack_query.aliases import (
+    ServerSideFilters,
+    ParsedFilterFunc,
+    OpenstackResourceObj,
+)
+
+
+class QueryRunner(OpenstackWrapperBase):
+    """
+    Base class for Runner classes.
+    Runner classes encapsulate running any openstacksdk commands
+    """
+
+    def __init__(self, connection_cls=OpenstackConnection):
+        OpenstackWrapperBase.__init__(self, connection_cls)
+
+    def run(
+        self,
+        cloud_account: str,
+        filter_func: Optional[ParsedFilterFunc] = None,
+        filter_kwargs: Optional[ServerSideFilters] = None,
+        from_subset: Optional[List[Any]] = None,
+        **kwargs
+    ) -> List[OpenstackResourceObj]:
+        """
+        Public method that runs the query by querying openstacksdk and then applying a filter function.
+        :param cloud_account: The account from the clouds configuration to use
+        :param filter_func: An Optional function that we can use to limit the results after querying openstacksdk
+        :param filter_kwargs: An Optional set of filter kwargs to limit the results by when querying openstacksdk
+        :param from_subset: A subset of openstack resources to run query on instead of querying openstacksdk
+        :param kwargs: An extra set of kwargs to pass to internal _run_query method that changes what/how the
+        openstacksdk query is run
+            - valid kwargs to _run_query is specific to the runner object - see docstrings for _run_query() on the
+            runner of interest.
+        """
+
+        force_filter_func_usage = False
+        with self._connection_cls(cloud_account) as conn:
+            if from_subset:
+                results_resource_objects = self._parse_subset(conn, from_subset)
+                force_filter_func_usage = True
+            else:
+                results_resource_objects = self._run_query(
+                    conn, filter_kwargs, **kwargs
+                )
+
+        if filter_kwargs is None or force_filter_func_usage:
+            results_resource_objects = self._apply_filter_func(
+                results_resource_objects,
+                filter_func,
+            )
```

Comment:
Can we simplify this flow a bit?

In hacky code:

```python
if from_subset and self._server_filters:
   raise ValueError("Cannot parse from a subset and server filter")

if not self._server_filters:
    self._server_filters = get_all_objects()

# Then we are always parsing a subset at this level
with ... as conn:
       subset = if from_subset if from_subset else self._run_query(self._server_filters)
        results_resource_objects = self._parse_subset(conn, self._client_filters, from_subset)
```

This would also drop the need to use `force_filter_func_usage` too